node/core/proposer/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+use futures::prelude::*;
+use futures::{lock::Mutex, select};
+use polkadot_node_subsystem::messages::{AllMessages, ProvisionableData, ProvisionerMessage};
+use polkadot_overseer::OverseerHandler;
+use polkadot_primitives::{
+	inclusion_inherent,
+	parachain::{ParachainHost, SignedAvailabilityBitfields},
+	Block, Header,
+};
+use sc_block_builder::{BlockBuilderApi, BlockBuilderProvider};
+use sp_api::{ApiExt, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_consensus::{Proposal, RecordProof};
+use sp_inherents::InherentData;
+use sp_runtime::traits::{DigestFor, HashFor};
+use sp_transaction_pool::TransactionPool;
+use std::{pin::Pin, sync::Arc, time};
+
+/// How long proposal can take before we give up and err out
+// REVIEW: this value is a SWAG; receptive to better ideas.
+const PROPOSE_TIMEOUT: core::time::Duration = core::time::Duration::from_secs(2);
```

Comment:
Better ideas for this value?

Reply:
what is a SWAG? Having a SWAGgy value sounds like a good thing to me 😎 

The principle of this value seems reasonable to me. I'd like to loop in @burdges because this is also part a larger conversation I'd be interested in having. 

Intuitively, there are a few places where we want to delay within the critical path of block production:
  * Waiting for availability pieces before issuing bitfields
  * Waiting for bitfields before proposing a set of bitfields and backable candidates
  * And this timeout: waiting for the `Provisioner` to provide the proposal. Essentially, the amount of time the authorship component waits for anything before just saying "fuck it" and authoring a block. Typically, the `Provisioner` will be gated on the above timeout or resolve near-instantly, but it is better practice not to rely on that and get stuck.

I'd be interested in exploring canonical values for these timeouts and hearing more information about how they are expected to interact with the slot timeouts for BABE or SASSAFRAS, given that we'd be releasing in the middle of the slot as opposed to the very beginning.

@andresilva is also doing some work on BABE where, as a secondary slot author, you also do another delay, and that might interact with any timeouts that we set.

Reply:
SWAG is an acronym I learned in the Army: shitty, wild-ass guess. It indicates relatively low confidence in the value.